File: quicksort/quicksort.py
import random
def quicksort(A):
    if len(A) <= 1: return A 
    i = random.randint(0, len(A)-1)
    # split to two groups
    less = []
    greater = [] 
    for element in A:
        if element == A[i]: continue 
        elif element < A[i]:
            less.append(element)
        else:
            greater.append(element)
    left = quicksort(less)
    right = quicksort(greater) 
    return left + [A[i]] + right 
    

File: quicksort/quicksort.ml
let rec split (l: int list) (pivot: int): int list * int list = 
    match l with 
    | [] -> ([],[])
    | x::l' -> let (less,greater) = split l' pivot
               in if (x < pivot) then (x::less, greater) else (less, x::greater)

let rec qsort (l: int list): int list = 
    match l with
    | [] -> []
    | x::[] -> l
    | x::l' -> let pivot = List.nth l 0 in 
               let (less, greater) = split l' pivot in 
               let ls_sorted = qsort less in 
               let gt_sorted = qsort greater in 
               List.append ls_sorted (x :: gt_sorted)

let _ = qsort([5;4;3;2;1])